fog/ledger/enclave/impl/src/key_image_store.rs

```diff
@@ -0,0 +1,192 @@
+///
+/// - The size in the OMAP is ValueSize which must be divisible by 8,
+/// - The user actually gives us a serialized protobuf
+/// - We use a wire format in the omap where value[0] = ValueSize - 1 -
+///   ciphertext.len(), ValueSize must be within 255 bytes of ciphertext.len().
+/// - When the lookup misses, we try to obliviously return a buffer of the
+///   normal size. We do this by remembering the ciphertext size byte of the
+///   last stored ciphertext.
+pub struct KeyImageStore<OSC: ORAMStorageCreator<StorageDataSize, StorageMetaSize>> {
+    /// Oblivious map to hold KeyImageStoreRecords
+    omap: Box<<ObliviousMapCreator<OSC> as OMapCreator<KeySize, ValueSize, McRng>>::Output>,
+
+    /// The logger object
+    logger: Logger,
+}
+
+impl<OSC: ORAMStorageCreator<StorageDataSize, StorageMetaSize>> KeyImageStore<OSC> {
+    pub fn new(desired_capacity: u64, logger: Logger) -> Self {
+        Self {
+            omap: Box::new(<ObliviousMapCreator<OSC> as OMapCreator<
+                KeySize,
+                ValueSize,
+                McRng,
+            >>::create(
+                desired_capacity, STASH_SIZE, McRng::default
+            )),
+            logger,
+        }
+    }
+
+    // add a key image containing block index and timestamp
+    pub fn add_record(
+        &mut self,
+        key_image: &KeyImage,
+        block_index: u64,
+        timestamp: u64,
+    ) -> Result<(), AddRecordsError> {
+        let mut value = A8Bytes::<ValueSize>::default();
+        let mut key = A8Bytes::<KeySize>::default(); // key used to add to the oram for key image
+        key.clone_from_slice(&key_image.as_ref());
+        // Flip the first byte of key image, when used as a key in the oblivious
+        // map. This is because we will use key image as a key in the map,
+        // but the map does not support all zeroes as a key. All zeroes is a
+        // valid curve point. But if we flip the first byte, it turns out that
+        // isn't a valid curve point, in the Ristretto group.
+        // So this prevents the OMAP_INVALID_KEY error path.
+        key[0] = !key[0];
+        // write block index data to  value[0..8] write the time stamp data to
+        // value[8..16]
+        value[0..8].clone_from_slice(&block_index.to_le_bytes());
+        value[8..16].clone_from_slice(&timestamp.to_le_bytes());
+        // Note: Passing true means we allow overwrite, which seems fine since
+        // the search_key value is not changing
+        let omap_result_code = self.omap.vartime_write(&key, &value, Choice::from(1));
+        if omap_result_code == OMAP_INVALID_KEY {
+            return Err(AddRecordsError::KeyRejected);
+        } else if omap_result_code == OMAP_OVERFLOW {
+            return Err(AddRecordsError::MapOverflow(
+                self.omap.len(),
+                self.omap.capacity(),
+            ));
+        } else if omap_result_code == OMAP_FOUND {
+            log::debug!(self.logger, "An omap key was found");
```

Comment:
nit: It would be good to have a slightly more informative log message -- maybe something like "A key image record was clobbered"? it is probably a bug if this happens